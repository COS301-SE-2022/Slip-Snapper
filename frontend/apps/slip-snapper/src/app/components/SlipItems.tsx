import { IonTitle, IonButton, IonCard, IonItem, IonAlert, IonCardHeader, IonCardTitle, IonLabel, IonSearchbar, IonCol, IonGrid, IonRow, IonToggle, IonDatetime, IonIcon, useIonToast } from '@ionic/react';
import React, { useEffect, useState } from 'react';
import { getAllSlips, deleteSlip } from '../../api/apiCall';
import '../theme/SlipItems.css';
import { calendarOutline, idCard } from 'ionicons/icons';



const SlipItems: React.FC = () => {
    let user = JSON.parse(localStorage.getItem('user')!)

    const [originalSlips, setOriginalSlips] = useState<any[]>([]);
    const [slipItems, setSlipItems] = useState<any[]>([]);
    const [present, dismiss] = useIonToast();


    useEffect(() => {
        if (user == null) {
            user = { id: 24 }
        }
        getAllSlips(user.id)
            .then(
                apiResponse => {
                    if (typeof (apiResponse.data) !== "string") {
                        orderSlips(apiResponse.data.slips)
                        setOriginalSlips(apiResponse.data.slips)
                        setSlipItems(apiResponse.data.slips)
                    }
                })
    }, []);
    const [deleteAlert, setDeleteAlert] = useState({
        state: false,
        name: '',
        id: 0,
    });

    const current = new Date();
    const [filterDates, setFilterDates] = useState({
        from: "",
        to: "",
    });
    return (
        <div>
            <IonItem>
                <IonTitle>All Receipts</IonTitle>
            </IonItem>
            <div className='wrapper'>
                <IonCard color="primary" className="search-bar">

                    <IonItem color='primary'>
                        <IonSearchbar color="tertiary" id='searchBar' onIonChange={filter} />
                    </IonItem>

                    <IonItem color='primary'>
                        <IonLabel>Total Filter</IonLabel>
                        <IonToggle color='secondary' />
                    </IonItem>

                    <IonItem color='primary'>
                        <IonLabel>Date Filter</IonLabel>
                        <IonToggle color='secondary' onIonChange={e => toggleDates(e.detail.checked)} />
                    </IonItem>

                    <div id='date-div' className='date-div' color="primary" >
                        <IonCardTitle className="date elem">From Date:
                            <IonItem className='date-item' color="tertiary">
                                <IonDatetime value={filterDates.from} displayFormat='DD/MM/YYYY' id={"fromDate"} />
                                <IonIcon icon={calendarOutline} slot="end" />
                            </IonItem>
                        </IonCardTitle>

                        <IonCardTitle className="date elem">To Date:
                            <IonItem className='date-item' color="tertiary">
                                <IonDatetime value={filterDates.to} displayFormat='DD/MM/YYYY' id={"toDate"} />
                                <IonIcon icon={calendarOutline} slot="end" />
                            </IonItem>
                        </IonCardTitle>

                        <IonItem color="primary" >
                            <IonButton color="secondary" slot="end" onClick={() => {
                                filter()
                            }}>Apply</IonButton>
                        </IonItem>

                    </div>
                </IonCard>

                <IonCard color="primary" className="receipts-table">
                    <IonCardHeader>
                        <IonCardTitle>Receipts</IonCardTitle>
                    </IonCardHeader>

                    {slipItems.map((item, index) => {
                        return (
                            <IonItem key={index} color="tertiary" id={"slipItem" + index}>
                                <IonLabel>
                                    {item.transactionDate.split('T')[0].replace(/-/gi, "/").split('/').reverse().join('/') + " - " + item.location}
                                </IonLabel>
                                {"Total: R" + item.total.toFixed(2)}
                                <IonButton routerLink="/editreceipt" id={item.id + "b"} color="secondary" slot="end" onClick={() => {
                                    localStorage.removeItem('editSlip')
                                    localStorage.setItem('editSlip', JSON.stringify(item))
                                }}>Edit</IonButton>
                                <IonButton
                                    onClick={() =>
                                        setDeleteAlert({
                                            state: true,
                                            name: item.location,
                                            id: item.id,
                                        })
                                    }
                                    fill="solid"
                                    slot="end"
                                    color="medium"
                                >
                                    Delete
                                </IonButton>
                                <IonAlert
                                    isOpen={deleteAlert.state}
                                    onDidDismiss={() =>
                                        setDeleteAlert({ state: false, name: '', id: 0 })
                                    }
                                    header="Confirm Delete"
                                    message="Are you sure you want to delete this slip?"
                                    buttons={[
                                        'Cancel',
                                        {
                                            text: 'Delete',
                                            cssClass: 'toasts',
                                            handler: async () => {
                                                await deleteSlip(deleteAlert.id)
                                                getAllSlips(user.id)
                                                    .then(
                                                        apiResponse => {
                                                            if (typeof (apiResponse.data) !== "string") {
                                                                setSlipItems(apiResponse.data.slips)
                                                            }
                                                        })
                                                setDeleteAlert({ state: false, name: '', id: 0 });
                                            },
                                        },
                                    ]}
                                />
                            </IonItem>
                        )
                    })}
                </IonCard>
            </div>
        </div>
    );

    function filter() {

        for (let i = 0; i < slipItems.length; i++) {
            const temp = document.getElementById("slipItem" + i)
            if (temp !== null)
                temp.style.display = "block";
        }

        const searchValue = document.getElementById("searchBar")?.getElementsByTagName("input")[0].value
        if (searchValue !== "") {
            searchFilter(searchValue)
        }

        if (document.getElementById("fromDate")?.getElementsByTagName("input")[0].value !== "" || document.getElementById("toDate")?.getElementsByTagName("input")[0].value !== "") {
            dateFilter()
        }

    }

    function searchFilter(searchText: string | undefined) {

        if (searchText !== undefined) {
            for (let i = 0; i < slipItems.length; i++) {
                if (!slipItems[i].location.toLowerCase().includes(searchText.toLowerCase())) {
                    const temp = document.getElementById("slipItem" + i)
                    if (temp !== null)
                        temp.style.display = "none";
                }
                else if (slipItems[i].transactionDate.split('T')[0].replace(/-/gi, "/").split('/').reverse().join('/').includes(searchText.toLowerCase())) {
                    const temp = document.getElementById("slipItem" + i)
                    if (temp !== null)
                        temp.style.display = "none";
                }
            }
        }
    }

    function toggleDates(state: any) {
        const temp = document.getElementById('date-div')
        if (state) {
            if (temp !== null)
                temp.style.display = "block";
        }

        if (!state) {
            if (temp !== null)
                temp.style.display = "none";

            setSlipItems(originalSlips)
        }
    }

    function dateFilter() {

        const fromDate = document.getElementById("fromDate")?.getElementsByTagName("input")[0].value.split('T')[0].replace(/-/gi, "/")
        const toDate = document.getElementById("toDate")?.getElementsByTagName("input")[0].value.split('T')[0].replace(/-/gi, "/")

        if (toDate !== "" && fromDate !== "" && toDate!==undefined&&fromDate!==undefined&&toDate<fromDate)
        {
            present('Please enter a valid Date interval.', 1200);
            return

        }

        if (fromDate !== "" && fromDate !== undefined) {
            for (let i = 0; i < originalSlips.length; i++) {
                if (fromDate > originalSlips[i].transactionDate) {
                    const temp = document.getElementById("slipItem" + i)
                    if (temp !== null)
                        temp.style.display = "none";
                }
            }
        }

        if (toDate !== "" && toDate !== undefined) {
            for (let i = 0; i < originalSlips.length; i++) {
                if (toDate < originalSlips[i].transactionDate) {
                    const temp = document.getElementById("slipItem" + i)
                    if (temp !== null)
                        temp.style.display = "none";
                }
            }
        }
        if (fromDate !== undefined && toDate !== undefined)
            setFilterDates({ from: fromDate, to: toDate })
    }

    function orderSlips(slips: any) {
        let temp: any;
        for (let i = 0; i < slips.length; i++) {
            for (let j = 1; j < (slips.length - i); j++) {
                if (slips[j - 1].transactionDate < slips[j].transactionDate) {
                    temp = slips[j - 1]
                    slips[j - 1] = slips[j]
                    slips[j] = temp;
                }
            }
        }
    }
};


export default SlipItems;

