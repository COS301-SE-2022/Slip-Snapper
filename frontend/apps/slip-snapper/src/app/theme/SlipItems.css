ul.no-bullets {
    list-style-type: none;
    line-height: 1.5;
}

.titles{
    --border-radius: 5px;
}

.details{
    --border-radius: 5px;
}

.wrapper {
    margin: 0.5rem;
    display: flex;
    flex-wrap: wrap;
    gap: 0.5rem;
    justify-content: center;
}

.receipts-items {
    vertical-align: middle;
    --border-radius: 10px;
    padding: 2px;
}
.receipts-table{
    flex:3;
    min-width: 20rem;
}

.search-bar{
    flex:2;
    min-width: 20rem;
}

.date-div {
    max-width: 200px;
    margin-left: 40px;
    display: none;
}

.date-item {
   max-width: 200px;
}

.totalSlider{
    margin-top: 30px;
    width: 70%;
    max-width: 85%;
    margin-left: 40px;
    display: none;
}
/* @media only screen and (max-width: 2200px) {
    .slip-item-col {
        min-width: 20%;
    }
}

*/