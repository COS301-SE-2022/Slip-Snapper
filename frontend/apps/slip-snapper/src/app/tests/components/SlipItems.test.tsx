import renderer from 'react-test-renderer';
import SlipItems from '../../components/SlipItems';

jest.mock('../../api/apiCall');
jest.mock("../../api/Session");

const renderTree = (tree: JSX.Element) => renderer.create(tree);
describe('<SlipItems>', () => {
  it('should render component', () => {
    expect(renderTree(<SlipItems 
    />).toJSON()).toMatchSnapshot();
  });
  
});