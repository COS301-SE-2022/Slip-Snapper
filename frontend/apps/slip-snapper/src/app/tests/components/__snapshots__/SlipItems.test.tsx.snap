// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SlipItems> should render component 1`] = `
<div>
  <ion-item
    onClick={[Function]}
  >
    <ion-title>
      All Receipts
    </ion-title>
  </ion-item>
  <ion-card
    color="primary"
    onClick={[Function]}
  >
    <ion-card-header>
      <ion-item
        color="primary"
        onClick={[Function]}
      >
        <ion-searchbar
          color="tertiary"
          id="searchBar"
        />
        <ion-fab
          horizontal="end"
        >
          <ion-fab-button
            color="secondary"
            onClick={[Function]}
            size="small"
          >
            <ion-icon
              src="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Filter</title><path stroke-linecap='round' stroke-linejoin='round' d='M32 144h448M112 256h288M208 368h96' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
            />
          </ion-fab-button>
        </ion-fab>
      </ion-item>
    </ion-card-header>
    <ion-item
      color="tertiary"
      id="emptySlips"
      onClick={[Function]}
    >
      Your scanned slips will be displayed here.
    </ion-item>
  </ion-card>
</div>
`;
